### PR TITLE
[Merged by Bors] - Update ATX handler to allow increasing numUnits again

### LINKED ISSUE #3963: Allow nodes to increase their claimed PoST size.
## Description
Nodes may increase their PoST size in the future and we should support this on a protocol level.

## Acceptance criteria
- Incoming ATXs are allowed to increase the claimed `numUnits` again.
- When an ATX is received that has a higher `NumUnits` value than the previous ATX of the same identity it may also contain a VRF Nonce that satisfies the increased difficulty.
  - Alternatively if it doesn't contain a new nonce the current needs to be re-validated to satisfy the increased difficulty
- The changed `numUnits` value becomes effective in the `TargetEpoch` of the ATX, when `numUnits` was lowered, but when `numUnits` is increased it becomes effective one epoch after `TargetEpoch`.
  - A changed VRFNonce always becomes effective in `TargetEpoch`.
  - see also these comments: https://github.com/spacemeshos/pm/issues/209#issuecomment-1377810006, https://github.com/spacemeshos/go-spacemesh/issues/3773#issuecomment-1339597893, https://github.com/spacemeshos/go-spacemesh/issues/3773#issuecomment-1374781236

## Implementation hints
- Introduce a new `EffectiveNumUnits` property for `VerifiedActivationTx` that returns the smaller `numUnits` value of the ATX and its `prevATX`.
  - This property might be persisted in the local DB for efficiency but should not be serialised when transmitting an ATX over the network.
- Use `EffectiveNumUnits` as input for `GetWeight` calculation of `VerifiedActivationTx`.

### LINKED ISSUE #3992: Add VRF nonce caching
### Description
At the moment hare queries the VRF nonce for any identity on every round. Since an identity cannot change their nonce within an epoch that is unnecessary and should be cached.

### Affected code
Everywhere nonces are fetched from the DB those queries should be cached based on the `NodeID` and `Epoch` such that a nonce for any identity only needs to be fetched at most once per epoch.

Related files (optionally with line numbers): 
- `hare/eligibility/oracle.go` and other places where nonces are fetched from the DB.

### Implementation hints:
- The nonce fetcher already uses `datastore.CachedDB`, which could be extended to cache nonces for identities on a per epoch basis.
- Consider re-using the malfeasance cache since it holds `nil` for all valid identities and could be used to store their VRF nonce.

### PR DESCRIPTION
**Pull request by fasmat**
## Motivation
Closes #3963 and #3992

## Changes
- Incoming ATXs with higher `numUnits` than their `prevAtx` are allowed again
- If `numUnits` is increased from one ATX to the next, re-validate the VRF for the identity
- VRFNonces are now cached when fetched from database: the nonce for an identity is fetched at most once per epoch from the DB and then served from cache
- `EffectiveNumUnits` has been introduced as a property of `ActivationTx` and `ActivationTxHeader`
  - it is the minimum of the `NumUnits` of the current and the previous ATX
  - it is persisted in a new column in the ATX table to not impact performance (otherwise we would always need to fetch 2 ATXs to determine its value)
  - it replaces `NumUnits` in the determination of the weight

## Test Plan
- Tests were added to check if increases in NumUnits are correctly validated by the ATX handler.
- Existing tests were updated to use set EffecitveNumUnits where relevant

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed

## DevOps Notes

- [x] This PR does not require configuration changes (e.g., environment variables, GitHub secrets, VM resources)
- [x] This PR does not affect public APIs
- [x] This PR does not rely on a new version of external services (PoET, elasticsearch, etc.)
- [x] This PR does not make changes to log messages (which monitoring infrastructure may rely on)

### CODE REVIEW

File: proposals/eligibility_validator_test.go [hunk 1]
```diff
@@ -293,48 +296,6 @@ func TestCheckEligibility_KeyMismatch(t *testing.T) {
 	require.False(t, eligible)
 }
 
-func TestCheckEligibility_ZeroTotalWeight(t *testing.T) {
```
Comment (fasmat) [inline proposals/eligibility_validator_test.go:296]: A node cannot have 0 weight an ATX with `NumUnits` 0 is not accepted as a valid ATX.

File: miner/oracle_test.go [hunk 2]
```diff
@@ -197,29 +198,3 @@ func TestOracle_OwnATXNotFound(t *testing.T) {
 	assert.Len(t, activeSet, 0)
 	assert.Len(t, proofs, 0)
 }
-
-func TestOracle_ZeroEpochWeight(t *testing.T) {
```
Comment (fasmat) [inline miner/oracle_test.go:201]: An ATX with `NumUnits` 0 is not a valid ATX, it will be refused by the ATX handler.

File: common/types/activation_tx_header.go [hunk 3]
```diff
@@ -10,6 +10,12 @@ type ActivationTxHeader struct {
 	// current epoch; a unit represents a configurable amount of data for PoST
 	NumUnits uint32
 
+	// EffectiveNumUnits is the minimum of this ATX's NumUnits and the previous ATX's NumUnits
+	// NumUnit decreases become effective immediately, while NumUnit increases become effective one epoch later
+	// This is because the increased PoST size only becomes effective after a PoET proof has
+	// been generated for it, which is published with the next epoch's ATX.
+	EffectiveNumUnits uint32
```
Comment (noamnelke) [inline common/types/activation_tx_header.go:17]: Isn't `ActivationTxHeader` the type that gets gossiped/synced? I don't think we should send this value to others, they should calculate it themselves.

File: common/types/nodeid.go [hunk 4]
```diff
@@ -9,8 +9,6 @@ import (
 	"github.com/spacemeshos/go-spacemesh/log"
 )
 
-//go:generate scalegen
-
```
Comment (noamnelke) [inline common/types/nodeid.go]: Seems unrelated 🤔

**Review by noamnelke — APPROVED** [review]

File: common/types/nodeid.go (hunk 4 above)
Comment (fasmat) [reply common/types/nodeid.go]: It breaks `make generate`; I usually run it before I push something for review. Maybe we should add it as a step to the CI?

It is not required, the only type in this file `NodeID` has a handwritten encode and decode method, but if we leave it `make generate` fails.

Comment (noamnelke) [reply common/types/nodeid.go]: Then keep the removal, but I agree that we should add this as a step to CI.

File: common/types/activation_tx_header.go (hunk 3 above)
Comment (fasmat) [reply common/types/activation_tx_header.go:17]: `ActivationTxHeader` is not broadcasted to the network. It is only used by the node to not need to fetch and deserialize the whole ATX from the database and for caching.

**Comment by fasmat** [comment]
bors try

File: common/types/activation_tx_header.go (hunk 3 above)
Comment (noamnelke) [reply common/types/activation_tx_header.go:17]: Ah, awesome then 👍🏻

**Comment by fasmat** [comment]
bors merge

File: common/types/activation_tx_header.go [hunk 5]
```diff
@@ -10,6 +10,12 @@ type ActivationTxHeader struct {
 	// current epoch; a unit represents a configurable amount of data for PoST
 	NumUnits uint32
```
Comment (countvonzero) [inline common/types/activation_tx_header.go:11]: why do the code still need this then?

Comment (countvonzero) [reply common/types/activation_tx_header.go:11]: NVM. dumb question.

File: activation/handler.go [hunk 6]
```diff
@@ -252,6 +263,11 @@ func (h *Handler) validateNonInitialAtx(ctx context.Context, atx *types.Activati
 		return fmt.Errorf("prevATX declared, but initial Post is included")
 	}
 
+	if prevAtx.NumUnits < atx.NumUnits {
+		atx.SetEffectiveNumUnits(prevAtx.NumUnits)
+	} else {
+		atx.SetEffectiveNumUnits(atx.NumUnits)
+	}
```
Comment (countvonzero) [inline activation/handler.go:270]: issue #3963 stated
> The changed numUnits value becomes effective in the TargetEpoch of the ATX, when numUnits was lowered, but when numUnits is increased it becomes effective one epoch after TargetEpoch.
A changed VRFNonce always becomes effective in TargetEpoch.

how is this achieved?

Comment (fasmat) [reply activation/handler.go:270]: The `EffectiveNumUnits` value is always the minimum of the `NumUnit` value of the ATX and its previous ATX. This `if` sets the value to the minimum of the two. The VRF Nonce becomes effective in the `TargetEpoch` because if `numUnits` was decreased the old VRF Nonce is still valid (still satisfies difficulty threshold) and if `numUnits` was increased the node will provide a new Nonce that is valid for both the lower and the higher `numUnits` value.

Example 1 - smesher decreases `numUnits`:

- Epoch 100: ATX published (`NumUnits` = 5, `VRFNonce` = 100)
- Epoch 101: ATX published (`NumUnits` = 3), used values `NumUnits` = 5, `VRFNonce` = 100
- Epoch 102: ATX published (`NumUnits` = 3), used values `NumUnits` = 3, `VRFNonce` = 100

The smesher decreases its `NumUnits` in epoch 101 from 5 to 3 which becomes effective immediately in the epoch after.

Example 2 - smesher increases `numUnits`:

- Epoch 100: ATX published (`NumUnits` = 5, `VRFNonce` = 100)
- Epoch 101: ATX published (`NumUnits` = 8, `VRFNonce` = 333), used values `NumUnits` = 5, `VRFNonce` = 100
- Epoch 102: ATX published (`NumUnits` = 8), used values `NumUnits` = 5, `VRFNonce` = 333
- Epoch 103: ATX published (`NumUnits` = 8), used values `NumUnits` = 8, `VRFNonce` = 333

The changed `VRFNonce` becomes effective immediately in the epoch after the ATX was published, but the increased `NumUnits` value becomes effective one epoch later. So that PoET can confirm the increase.

**Comment by fasmat** [comment]
bors merge
